Run vale linter only if it is installed, skip otherwise

OrkoHunter opened the pull request:
Vale linter was added to the project in #2631 as a language linter. It runs when Markdown files are changed.

**Problem**

A developer is not be able to commit if vale command fails to run. In case when the user does not have vale installed, the error message is unhelpful, thus causing a bad contributor experience. See relevant discussion https://discordapp.com/channels/687207715902193673/687994765559463940/765132734195564574

**Proposed**

- Run `scripts/check-docs-quality.js` as a `lint-staged` script, instead of raw `vale` CLI.
- The script passes silently if vale is not installed. (CI will catch the lint issues on PR nevertheless.)
- If vale is installed, the developer will be prompted to fix language lint issues.

OrkoHunter commented on package.json:
```diff
@@ -41,6 +41,7 @@
     "@changesets/cli": "^2.11.0",
     "@spotify/eslint-config-oss": "^1.0.1",
     "@spotify/prettier-config": "^8.0.0",
+    "command-exists": "^1.2.9",
```
Looks safe to add plus it is cross platform. [A good ~150 lines of helper code](https://github.com/mathisonian/command-exists/blob/master/lib/command-exists.js).

Feel free to suggest any alternative.

andrewthauer commented on scripts/check-docs-quality.js:
```diff
@@ -16,53 +16,67 @@
+      const validMDFilesCommand = `${listFilesTrackedByGit} | .\\node_modules\\.bin\\shx grep ".md"`;
+      try {
+        // get list of all md files except in directories of gitignore.
+        let filesToLint = execSync(validMDFilesCommand, {
```
As suggestion, this could leverage the fact that `lint-staged` passes the files to line to the command it invokes. Therefore this subshell could be avoided if there are files in the `argv` array.

For example when `node ./scripts/check-docs-quality` is invoked by lint-staged, it should result in `node ./scripts/check-docs-quality file1.md path/to/file2.md file3.md`

OrkoHunter replied:
Valid suggestion @andrewthauer ! Let me explore this a bit. And TIL `lint-staged` passes the list of files to the command. :)

Marvin9 replied:
We could add the flag which tracks if this script was called through ```lint-staged``` or manually. If the script is called through ```lint-staged```, we have all valid files in ```argv```. @OrkoHunter Sorry for missing out this case but I can solve it.

OrkoHunter replied:
Hey @Marvin9, what I am thinking is that if the script receives any arguments, it will lint only those files. If the script receives no arguments, it will lint all the (md) files. And no need to apologise at all. :)

Marvin9 replied:
That's also fine & cleaner solution. In that case[provided files eg. length of argv > 2] we can use spawnSync.

```js
spawnSync('vale', process.argv.slice(2), inheritStdIo);
```

Rugvip commented on scripts/check-docs-quality.js:
```diff
@@ -16,53 +16,67 @@
-      stdio: ['ignore', 'pipe', 'inherit'],
-    });
+// Proceed with the script only if Vale linter is installed. Limit the friction and surprises caused by the script.
+commandExists('vale')
```
Probably makes sense to keep the `lint-staged` logic behind a flag, or the other way around. We wanna make sure that `vale` is always present in CI, both for us and whoever might reuse this script.

OrkoHunter replied:
Makes sense.

The script can check if `CI` env variable is set to true. If yes, the script will make sure `vale` is installed. If false, the script assumes to be run by `lint-staged` and does not cause errors if `vale` is not installed.

Ref: `CI` is always set to true on GitHub CI https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables

OrkoHunter replied:
Done!

OrkoHunter commented:
@Rugvip @andrewthauer @Marvin9 Thanks for your comments! Addressed the feedback about using files provided by `lint-staged` and ensuring the correct behaviour on CI vs local development. Please take a look again, thanks!

Marvin9 commented on scripts/check-docs-quality.js:
```diff
@@ -15,54 +14,76 @@
+    stdio: ['ignore', 'pipe', 'inherit'],
+  })
+    .toString()
+    .split('\n');
```
Are these relative paths working in windows? Just wants to confirm.

OrkoHunter replied:
I am not sure to be honest. If you can test this branch on Windows, it will be amazing!

Although, our contributors (on any OS) probably would not encounter this line since `lint-staged` will provide the absolute path to those files. 💯

freben reviewed (approved):
I'm happy with and about this. But I'm not sure why codecov would have an opinion.

OrkoHunter commented:
@freben Thanks! We have set codecov threshold to 0% for our project and this PR adds 0.16% of code without tests. :P 
It's fine to ignore, yes. That ❌ doesn't look so good, true. :/
